server/apps/migration/core-data-jpa-to-pg/src/main/java/org/apache/james/JpaToPgCoreDataMigration.java

```diff
@@ -0,0 +1,455 @@
+        migration.start();
+    }
+
+    static class DomainMigration {
+        private static final Logger LOGGER = LoggerFactory.getLogger(DomainMigration.class);
+
+        private final JPADomainList jpaDomainList;
+        private final PostgresDomainList pgDomainList;
+
+        @Inject
+        DomainMigration(
+                JPADomainList jpaDomainList,
+                PostgresDomainList pgDomainList
+        ) {
+            this.jpaDomainList = jpaDomainList;
+            this.pgDomainList = pgDomainList;
+        }
+
+        void doMigration() {
+            LOGGER.info("Start domains migration");
+            try {
+                jpaDomainList.getDomains().forEach(domain -> {
+                            try {
+                                pgDomainList.addDomain(domain);
+                            } catch (DomainListException e) {
+                                if (!e.getMessage().contains("already exists.")) {
+                                    LOGGER.warn("Failed to migrate domain {}", domain, e);
+                                }
+                            }
+                        }
+                );
+
+            } catch (DomainListException e) {
+                throw new RuntimeException("Unable to migrate domains, aborting processing", e);
+            }
+            LOGGER.info("Domains migration completed");
+        }
+    }
+
+    static class UsersMigration {
+        private static final Logger LOGGER = LoggerFactory.getLogger(UsersMigration.class);
+        private final Algorithm algorithm;
+        private final JPAUsersDAO jpaUsersDAO;
+        private final PostgresUsersDAO postgresUsersDAO;
+
+
+        @Inject
+        UsersMigration(
+                JPAUsersDAO jpaUsersDAO,
+                PostgresUsersDAO postgresUsersDAO,
+                PostgresUsersRepositoryConfiguration postgresUsersRepositoryConfiguration
+        ) {
+            this.algorithm = postgresUsersRepositoryConfiguration.getPreferredAlgorithm();
+            this.jpaUsersDAO = jpaUsersDAO;
+            this.postgresUsersDAO = postgresUsersDAO;
+        }
+
+        void doMigration() {
+            LOGGER.info("Start users migration");
+            try {
+                jpaUsersDAO.list().forEachRemaining(username -> {
+                            try {
+                                jpaUsersDAO.getUserByName(username).ifPresent(user -> {
+                                    var jpaUser = (JPAUser) user;
+                                    var pgUser = new DefaultUser(
+                                            username,
+                                            jpaUser.getPasswordHash(),
+                                            jpaUser.getAlgorithm(),
+                                            algorithm
+                                    );
+                                    try {
+                                        postgresUsersDAO.addUser(username, UUID.randomUUID().toString());
```

Comment:
Don't we have a way to preserve user password? By outputing the same hash into PG?

Reply:
the migration for users uses 2 steps : 
1. add the user with a random password
2. update the user which restores its password and hashing algorithm 

I did it this way to minimize the impact on the repository APIs